2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,234 @@
+
+One way to derive names is to consider other things in the same category as the current implementation and then to name that category. What other things might you do to a name to present it? A simple matrix helps:
+
+| Person name                  | ???               |
+|------------------------------|-------------------|
+| Remove extraneous whitespace | `trim_whitespace` |
+| Present last name first      | `last_first`      |
+| Capitalize words             | `capitalize`      |
+| Prepend salutation           | `add_salutation`  |
+
+The verb `present` is a viable candidate for the missing column heading in the matrix above. Remember that the name of the module is `PersonPresenter` so naming the function present would read like an echo chamber: `PersonPresenter.present_name`.
+
+There's another issue with the name `present`. From the internal perspective of the `PersonPresenter` module, everything is about presentation. But naming a public function demands that you consider its purpose from an outside perspective. Function callers want the presenter to do something specific to a person's name. They want it to be formatted. Context matters. `format_name` is a better choice. 
+
+```elixir
+defmodule PersonPresenter do
+  def format_name(name)
+    name
+    |> String.split()
+    |> Enum.reverse
+    |> Enum.join(", ")
+  end
+end
+```
+
+Context and perspective also dictate that the function be named not just `format`, but `format_name` Why include `_name` when the function will accept `name` as a parameter? Despite the fact that it, too, creates a bit of an echo chamber, it contains the right amount of information from the caller's perspective. Right away you know what this does: `PersonPresenter.format_name("  Jane  Doe ")`. You are not forced to open up a file and understand implementation details. `format_name` also leaves open the possibility of something like `format_address`.
+
+If the module were named `PersonFormatter`, it might make more sense to name the function something else, perhaps simply `name`. Naming the module is outside the purview of this article, but will be addressed in a followup article.
+
+Diagram 2: Function names abstraction ladder
+
+<img src="images/function-names-ladder.svg" width="100%" alt="Function Names Abstraction Ladder" />
+
+Should you need to change how the name is formatted in response to upstream data changes or new application requirements, the impact of the implementation will be constrained to the internals of the function. The name will continue to be appropriate and no dependents will be forced to change.
+
+In terms of the Open Closed Principle, the code is more open to the change. For example, a new requirement to add a salutation would only necessitate changes to implementation internals, no API-breaking name change would be necessary. By increasing the level of abstraction, you have created an approprate and resilient function name.
```

Comment:
"appropriate" is misspelled in the last sentence (as "approprate")

Reply:
Good catch! Resolved in 953c87eedef5e23f55300d93cb56f52f381f0f6c.